### PR TITLE
Retrieve transactions by search index - Closes #247

### LINKED ISSUE #247: Retrieve transactions by search index
### Description

Make the version 2 API able to retrieve data by various criteria.

Make sure that additional conditions are met according to the specification below.

Validation rules, intervals and type checking need to be performed at the Gateway level.

### Motivation

- Random access and search by those properties is required by the UI

### Acceptance Criteria

- Version 2 API supports additional search parameters

#### Endpoints

- HTTP `/api/v2/transactions`
- RPC `get.transactions`

#### Request parameters

Parameter | Type | Validation | Default | Comment
-- | -- | -- | -- | --
transactionId | String | `/^([1-9]\|[A-Fa-f0-9]){1,64}$/` | *(empty)* |  
moduleAssetId | String | `ModuleId:AssetId/[0-9]+:[0-9]+/` | *(empty)* | Transfer transaction: moduleID = 2,assetID = 0
moduleAssetName | String | `ModuleName:AssetName/[a-z]+:[a-z]+/` | *(empty)* | Transfer transaction: moduleName = token, assetName = transfer
senderAddress | String | `/^lsk([a-hjkm-z]\|[2-9]){38}$//^[1-9]\d{0,19}[L\|l]$/` | *(empty)* |  
senderPublicKey | String | `/^([A-Fa-f0-9]{2}){32}$/` | *(empty)* |  
senderUsername | String | `/^[a-z0-9!@$&_.]{1,20}$/` | *(empty)* |  
recipientAddress | String | `/^lsk([a-hjkm-z]\|[2-9]){38}$//^[1-9]\d{0,19}[L\|l]$/` | *(empty)* |  
recipientPublicKey | String | `/^([A-Fa-f0-9]{2}){32}$/` | *(empty)* |  
recipientUsername | String | `/^[a-z0-9!@$&_.]{1,20}$/` | *(empty)* |  
data | String |   | *(empty)* | Wildcard search ie. “gene*” -> “genesis”
amount | String |   | *(empty)* | Can be expressed as interval ie. `100000:200000`
timestamp | String |   | *(empty)* | Can be expressed as interval ie. `100000:200000`
includePending | Boolean |   | false |  
nonce | String | `/^\d+$/` | *(empty)* | In conjunction with senderAddress
limit | Number | `<1;100>` | 10 |  
offset | Number | `<0;+Inf>` | 0 |  
sort | String | `[“amount:asc”, “amount:desc”, “timestamp:asc”, “timestamp:desc”]` | timestamp:desc |  

#### Response example

200 OK

Make the version 2 API able to retrieve data by those criteria.

```jsonc
{
  "data": [
    {
      "id": "222675625422353767",
      "operationId": "2:0",
      "operationName": "token:transfer",
      "fee": "1000000",
      "nonce": "0",
      "block": { // optional
        "id": "6258354802676165798",
        "height": 8350681,
        "timestamp": 28227090,
      },
      "sender": {
        "address": "lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu",
        "publicKey": "2ca9a7...c23079",
        "username": "genesis_51",
      },
      "signatures": [ "72c9b2...36c60a" ],
      "confirmations": 0,
      "asset": {     // Depends on operation
        "amount": "150000000",
        "recipient": {
          "address": "lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu",
          "publicKey": "2ca9a7...c23079",
          "username": "genesis_49",
        },
        "data": "message"
      },
      "relays": 0,
      "isPending": false
    }
  ],
  "meta": {
    "count": 100,
    "offset": 25,
    "total": 43749
  },
  "links": {}
}

```

400 Bad Request

```jsonc
{
  "error": true,
  "message": "Unknown input parameter(s): <param_name>"
}
```

404 Not Found

```jsonc
{
  "error": true,
  "message": "Transaction <transaction_id> not found."
}

```

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #247 

### How was it solved?
- [x] Version 2 API supports additional search parameters
- [x] Search by given properties works
  - [x] id
  - [x] address
  - [x] moduleAssetId
  - [x] moduleAssetName
  - [x] blockId
  - [x] block height
  - [x] sender: address, public key and username
  - [x] recipient: address, public key and username
  - [x] amount as range
  - [x] timestamp as range
  - [x] nonce with sender address
  - [x]  limit, offset and sort
  - [x] search (/api/v2/transactions?search=(searchString))
  - [x] data (wildcard search on asset.data)

### How was it tested?
Local

### CODE REVIEW

File: services/core/shared/core/helpers/accountSearch.js [hunk 1]
```diff
@@ -13,32 +13,34 @@
...
+
+const nop = () => {};
 
 module.exports = {
-    getAddressByPublickKey,
-    getAddressBySecPublickKey,
-    getAddressByusername,
+    getAddressByPublickKey: nop,
+    getAddressBySecPublickKey: nop,
+    getAddressByusername: nop,
```
Comment (sameersubudhi) [inline services/core/shared/core/helpers/accountSearch.js:45]: Please ensure camel casing for the method names. And, if we do not need any of the methods here, please consider removing the file.

```suggestion
    getAddressByUsername: nop,
```

File: services/core/shared/core/compat/sdk_v2/blocks.js [hunk 2]
```diff
@@ -50,3 +50,80 @@ const getBlocks = async params => {
 };
 
 module.exports = { getBlocks };
+
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v2/blocks.js:53]: Can we please clean the commented lines of code, if unnecessary?

File: services/core/shared/core/compat/sdk_v5/accounts.js [hunk 3]
```diff
@@ -44,19 +49,31 @@ const confirmPublicKey = async publicKey => {
...
-	const accountsDB = await knex('accounts');
+	const accountsDB = await getAccountsIndex();
 	const account = await accountsDB.find({ publicKey });
 	return account;
 };
 
 const getIndexedAccountInfo = async params => {
-	const accountsDB = await knex('accounts');
+	const accountsDB = await getAccountsIndex();
+	const account = await accountsDB.find(params);
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/accounts.js:65]: Please consider unwrapping `account` at source and returning only the object. This change would also require you to update at places wherever the method has been invoked.

Same could be said for `getIndexedAccountByPublicKey`. But if possible, would appreciate replacing it completely with `getIndexedAccountInfo`. Seems a little redundant. Alternatively, remove `getIndexedAccountInfo` and replace it couple more new methods: `getIndexedAccountByAddress` and `getIndexedAccountByUsername`.

```suggestion
	const [account] = await accountsDB.find(params);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Please find my remarks.

File: services/core/shared/core/helpers/accountSearch.js (hunk 1 above)
Comment (nagdahimanshu) [reply services/core/shared/core/helpers/accountSearch.js]: Removed the given file.

File: services/core/shared/core/compat/sdk_v5/accounts.js (hunk 3 above)
Comment (nagdahimanshu) [reply services/core/shared/core/compat/sdk_v5/accounts.js]: I removed `getIndexedAccountByPublicKey` and replaced it with updated `getIndexedAccountInfo`.

File: services/core/shared/core/compat/sdk_v5/accounts.js [hunk 4]
```diff
@@ -43,20 +48,20 @@ const confirmPublicKey = async publicKey => {
...
+	const accountsDB = await getAccountsIndex();
+	const params = {
+		search: {
+			property: 'username',
+			pattern: searchString,
+		},
+	};
 	const account = await accountsDB.find(params);
 	return account;
 };
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/accounts.js:67]: Please consider renaming it to `getAccountsByUsernameSearch`.
```suggestion
const getAccountsByUsernameSearch = async searchString => {
	const accountsDB = await getAccountsIndex();
	const params = {
		search: {
			property: 'username',
			pattern: searchString,
		},
	};
	const account = await accountsDB.find(params);
	return account;
};
```

Or better make it more generic, with something similar to:

The `searchProp` and `searchString` can be wrapped into a single object as well.
```
{
    property: 'searchProp'
    pattern: 'searchString',
}
```
```suggestion
const getAccountsBySearch = async searchProp, searchString => {
	const accountsDB = await getAccountsIndex();
	const params = {
		search: {
			property: searchProp,
			pattern: searchString,
		},
	};
	const account = await accountsDB.find(params);
	return account;
};
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/core/shared/core/compat/sdk_v5/accounts.js (hunk 4 above)
Comment (nagdahimanshu) [reply services/core/shared/core/compat/sdk_v5/accounts.js]: Done!

**Review by sameersubudhi — APPROVED** [review]
LGTM!

**Review by MichalTuleja — CHANGES_REQUESTED** [review]
Please consider adding index trigger by recipient address. Make sure that the publicKey is not overwritten during updates.
Update: Please do it in #250

**Review by MichalTuleja — APPROVED** [review]

**Pull request merged**